File: services/orbital-engine/orbital_engine/api/audit.py
"""Audit query endpoint (Stage 5): ADMIN-only, read-only view of the log."""

from __future__ import annotations

from datetime import datetime
from typing import Annotated, Any

from fastapi import APIRouter, Depends, Query
from pydantic import BaseModel

from orbital_engine.security.attributes import Action, DataDomain, Subject
from orbital_engine.security.audit import query_audit
from orbital_engine.security.enforce import requires

router = APIRouter(tags=["audit"])


class AuditEntry(BaseModel):
    id: int
    ts: datetime
    subject: str | None = None
    role: str | None = None
    service: str | None = None
    source_ip: str | None = None
    domain: str
    action: str
    decision: str
    code: str
    reason: str
    method: str | None = None
    path: str | None = None


class AuditPage(BaseModel):
    """One page of audit rows plus the total matching count (for page-count calc)."""

    rows: list[AuditEntry]
    total: int


@router.get("/audit", response_model=AuditPage, summary="Query the audit log (ADMIN)")
async def get_audit(
    subject: str | None = Query(default=None, description="Substring match on subject"),
    path: str | None = Query(default=None, description="Substring match on request path"),
    source_ip: str | None = Query(default=None, description="Substring match on source IP"),
    decision: Annotated[list[str] | None, Query(description="permit / deny (OR)")] = None,
    role: Annotated[list[str] | None, Query(description="Filter by role (OR)")] = None,
    domain: Annotated[list[str] | None, Query(description="Filter by data domain (OR)")] = None,
    action: Annotated[list[str] | None, Query(description="Filter by action (OR)")] = None,
    ts_from: Annotated[datetime | None, Query(description="Earliest ts (inclusive)")] = None,
    ts_to: Annotated[datetime | None, Query(description="Latest ts (inclusive)")] = None,
    limit: int = Query(default=100, ge=1, le=1000, description="Page size"),
    offset: int = Query(default=0, ge=0, description="Rows to skip (page * size)"),
    _admin: Subject = Depends(requires(DataDomain.AUDIT, Action.READ)),
) -> dict[str, Any]:
    return await query_audit(
        subject=subject,
        path=path,
        source_ip=source_ip,
        decision=decision,
        role=role,
        domain=domain,
        action=action,
        ts_from=ts_from,
        ts_to=ts_to,
        limit=limit,
        offset=offset,
    )

File: apps/web/src/lib/api/audit.ts
"use client";

// Typed contract between the Audit Log table and the data-fetching layer.
//
// The engine paginates and filters server-side (see orbital_engine/api/audit.py):
// `AuditQuery` is what the table asks for, `AuditPage` is what the server returns.
// `buildAuditParams` is the single place the query is mapped to wire params, so the
// table never hand-builds query strings and the param names stay in one spot.

import { useApiQuery } from "./useApiQuery";
import type { QueryParams } from "./apiClient";

/** One audit row, mirroring the engine's `AuditEntry` model. */
export interface AuditEntry {
  id: number;
  ts: string;
  subject: string | null;
  role: string | null;
  service: string | null;
  source_ip: string | null;
  domain: string;
  action: string;
  decision: string;
  code: string;
  reason: string;
  method: string | null;
  path: string | null;
}

/** A page of rows plus the total matching count (drives page-count math). */
export interface AuditPage {
  rows: AuditEntry[];
  total: number;
}

/**
 * Everything the server needs to answer one table render: the pagination window
 * plus the active filters. Substring filters (`subject`, `path`) are contains
 * matches; the array filters are inclusive-OR multi-selects. Omitted/empty fields
 * mean "no filter".
 */
export interface AuditQuery {
  pageIndex: number;
  pageSize: number;
  subject?: string;
  path?: string;
  source_ip?: string;
  decision?: string[];
  role?: string[];
  domain?: string[];
  action?: string[];
  /** Inclusive timestamp range bounds, ISO 8601 strings. */
  tsFrom?: string;
  tsTo?: string;
}

/** Map a typed query to wire params (offset = page * size; arrays repeat). */
export function buildAuditParams(query: AuditQuery): QueryParams {
  return {
    limit: query.pageSize,
    offset: query.pageIndex * query.pageSize,
    subject: query.subject,
    path: query.path,
    source_ip: query.source_ip,
    decision: query.decision,
    role: query.role,
    domain: query.domain,
    action: query.action,
    ts_from: query.tsFrom,
    ts_to: query.tsTo,
  };
}

/**
 * Fetch one page of the audit log. `keepPreviousData` keeps the prior page visible
 * while the next request is in flight, so paging/filtering shows a loading hint
 * (via `isFetching`) instead of flashing empty. The full query is part of the key,
 * so any pagination or filter change refetches automatically.
 */
export function useAuditLog(query: AuditQuery) {
  return useApiQuery<AuditPage>({
    queryKey: ["audit", "log", query],
    url: "/api/audit",
    options: { params: buildAuditParams(query), keepPreviousData: true, retry: 0 },
  });
}
